helix-core/src/test/java/org/apache/helix/integration/rebalancer/CrushRebalancers/TestCrushAutoRebalanceNonRack.java

```diff
@@ -278,6 +280,17 @@ public void testLackEnoughInstances(String rebalanceStrategyName, String rebalan
         return !InstanceValidationUtil.isEnabled(helixDataAccessor, p.getInstanceName())
             && !InstanceValidationUtil.isAlive(helixDataAccessor, p.getInstanceName());
       }, TestHelper.WAIT_DURATION), "Instance should be disabled and offline");
+
+      // There is still a possibility that controller may see the liveinstance before the instance
+      // is stopped, even from admin tool perspective the node is not alive and disabled. Thus sending
+      // messages before or during instance getting dropped.
+      Assert.assertTrue(TestHelper.verify(() -> {
+        PropertyKey.Builder propertyKeyBuilder = helixDataAccessor.keyBuilder();
+        Message messages =
+            helixDataAccessor.getProperty(propertyKeyBuilder.messages(p.getInstanceName()));
+        return messages == null;
+      }, TestHelper.WAIT_DURATION));
```

Comment:
As synced offline, checking messages empty is not good enough as messages might be sent before dropping instance and hence the test will be flaky again. We should ensure the state is stable after instance is disabled, so messages won't be sent to the instance. Then dropping instance is safe.

Reply:
Right. In fact, the more fundamental issue seems to be liek this:
Dropping instance from test or admin tool vs controller adding messages to the same instance are async operations. The have race condition fundamentally. This also happens in production. This is a design flaw in Helix IMO. 
That said, to stabilize the test, we can disable controller and enable it again afterward. This is workaround for test, not production though.